### DISCUSSION TITLE
[Help] ES Modules for Dummies?

### DISCUSSION THREAD

**Discussion by JonWallsten** [webpack/webpack#14802] [Q&A]
I'm trying to wrap my head around the new ECMAScript modules.
When we upgraded Angular to v13 we were forced to handle ESM in our build chain.
Since we now kind of have support for it would be nice to convert our own internal libraries into ESM as well.
But there seem to be many different parts involved and it's hard to understand who's doing what.

We have `compilerOption.module` in `tsconfig.json`:
```
{
  "compilerOptions": {
    "module": "ES2020"
  }
}
```
Then we have `output.libraryTarget` in `webpack.config.ts`:
```
const config = {
    output: {
        libraryTarget: 'module',
    },
    experiments: {
        outputModule: true
    }
}
```

We also need `{ "type": "module" }` in package.json, correct?

To make things more complicated we have the `webpack.config` in TypeScript, which involves `ts-node`.

As I understand, when you use `{ "type": "module" }` the resolving of modules does something different? 
I can no longer use TypeScript for the config file:
```
[webpack-cli] Failed to load 'C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts' config
[webpack-cli] TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts
```
To fix that I add `--loader ts-node/esm` like this:
`node --loader ts-node/esm ../../node_modules/webpack/bin/webpack.js --config config/webpack.dev.ts`

That leads to the next error:
```
[webpack-cli] Failed to load 'C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts' config
[webpack-cli] CustomError: Cannot find module 'C:\Users\user\repo\packages\web-lib-core\config\helpers' imported from C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts
```
I import helpers from a local TS file:

`helpers.ts` includes some simple tooling for paths:
```ts
import { join, resolve } from 'path';
import { realpathSync } from 'fs';

const ROOT_PATH = resolve(__dirname, '..');
export const rootPath = (...paths: string[]): string => join(ROOT_PATH, ...paths);
export const realRootPath = (...paths: string[]) => realpathSync(rootPath(...paths));

export const include = [
    ...includeBase,
    rootPath('src'),
    rootPath('typings')
];
```

After reading this discussion: https://github.com/webpack/webpack-cli/issues/2458
I added this to my `tsconfig.json`:
```
{
  "ts-node": {
    "compilerOptions": { "module": "commonjs" }
  }
}
```

This gives me the next error:
```
[webpack-cli] Failed to load 'C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts' config
[webpack-cli] ReferenceError: exports is not defined in ES module scope
    at file:///C:/Users/user/repo/packages/web-lib-core/config/webpack.dev.ts:2:23
```

Using `--experimental-specifier-resolution=node` from https://github.com/TypeStrong/ts-node/issues/1007#issue-598417180 gives me this instead:

```
[webpack-cli] Failed to load 'C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts' config
[webpack-cli] file:///C:/Users/user/repo/packages/web-lib-core/config/webpack.base.ts:1
import { DefinePlugin } from 'webpack';
         ^^^^^^^^^^^^
SyntaxError: Named export 'DefinePlugin' not found. The requested module 'webpack' is a CommonJS module, which may not support all 
module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'webpack';
const { DefinePlugin } = pkg;
```

Removing that flag and instead using this approach:
```
import webpackMerge from 'webpack-merge';
import * as helpers from './helpers.js'; // .js added
import commonConfig from './webpack.base.js'; // .js added
```

Yields the same error as with the flag.

When following the instructions in the error message (which was a hassle since we deconstruct almost all imports): 
```
[webpack-cli] Failed to load 'C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts' config
[webpack-cli] C:\Users\user\repo\build-tools\helpers.ts:1
import * as path from 'path';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

And that's were I am today.
What is the preferred approach when working with Webpack, ESM and TypeScript? I know it's all still experimental (https://github.com/webpack/webpack-cli/issues/2458#issuecomment-811448600) but it would be nice to stay on top of it.
Should I just let it go until all the tooling is more mature?
Should I move the helpers to it's own library and compile it?
Should I compile the config files and before the build?
Is there any other approach I'm not aware of?

I'm also a bit curious about the relationship between `libraryTarget: 'module'` and `compilerOptions.module: 'ESXXXX'`. I didn't find that much info about `libraryTarget: 'module'` in the docs. But I guess everything is tied together?
Any explanation or help otherwise in appreciated!
@alexander-akait @sokra

**Comment by alexander-akait** [comment]
> We also need { "type": "module" } in package.json, correct?

What is output of your source code? ESM? Here many information https://www.typescriptlang.org/docs/handbook/esm-node.html, if you don't want to read deeply, you can use https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

But it is not related to webpack, because we handle and `import` and `require`

Do you build library? If no, you don't need `libraryTarget`

> [webpack-cli] CustomError: Cannot find module 'C:\Users\user\repo\packages\web-lib-core\config\helpers' imported from C:\Users\user\repo\packages\web-lib-core\config\webpack.dev.ts

check your import, you should have `.js` extension at the end

> const ROOT_PATH = resolve(__dirname, '..');

no `__dirname` in ESM

What is the main problem? You can't run webpack?

**Comment by JonWallsten** [comment]
@alexander-akait: We build libraries in pure TypeScript and Angular applications on top of that. But this is only for the libraries. 

I will read the links you posted. Thanks!

I didn't post this as a bug, but rather a question to get some answers from people with a better understanding then me right now. 
So I'm not stating that anything is webpacks fault.

My main problem right now is that I can't build the library att all. And following all the workarounds in the threads I linked feel more like a hack then the correct approach. Is it possible to run the build (with config files in TypeScript) as before but output the library as a module?

The helpers are not exported in the module, they are just used during build. So `__dirname` should be fine?

**Reply by alexander-akait** [reply]
To be honest, I cannot understand the problem...

Do you have problems with bundling or problems with run webpack using ts under esm?

If you want a quick solution just setup minimum example repo (just `package.json`/`webpack.config.ts` and some code with problems places).

To be honest, I have done and shown this many times already, most likely a small mistake in some configuration

**Reply by JonWallsten** [reply]
@alexander-akait I got it to build now. My main issue was that I did not put the config/helper files under exclude in `tsconfig.json`, so it became part of the compilation and that's why I got all the errors. 
I can't however consume the library in my other library.
```
ERROR in ./src/components/table/classes/model.table.cell.class.ts 1:0-175
Module not found: Error: Can't resolve '@oas/web-lib-core' in 'C:\Users\user\repo\packages\web-lib-common\src\components\table\classes'
 @ ./src/components/table/classes/model.table.row.class.ts 3:0-61 39:25-42 70:35-52
 @ ./src/index.ts 4:0-84 4:0-84
```

![image](https://user-images.githubusercontent.com/5993909/143181610-331f01e3-b22e-40ba-bbdd-9fbadfefa0b6.png)

I'll set up a repro and maybe you can have a quick look when you have time! thanks! :)

**Reply by JonWallsten** [reply]
So I finally got "externals" to work with our own libraries. So everything is built without errors.
However I get this runtime issue:
![image](https://user-images.githubusercontent.com/5993909/143187268-4c6634a5-747c-4196-99bf-5444eca7895c.png)

**Reply by JonWallsten** [reply]
![image](https://user-images.githubusercontent.com/5993909/143187354-cbff6a14-6e4d-42c9-8ad3-9e436ccc614a.png)

**Reply by JonWallsten** [reply]
Still working of that repro. We have a quite complex setup.

**Reply by JonWallsten** [reply]
@alexander-akait: I have made a repo now. Please have a look wherever you have time. You already have access to the repo: https://github.com/JonWallsten/monorepo-new/tree/esm

In short:
`web-lib-core`/`web-lib-common` have been converted to modules.
`web-lib-angular` is packaged with `ng-packagr` as a module. This is consuming `web-lib-core`
`web-app-edit` is consuming `web-lib-core` and `web-lib-angular` 

The error is this:
![image](https://user-images.githubusercontent.com/5993909/143212900-8150565f-dab2-40e5-bb22-55aa5c5c5721.png)
![image](https://user-images.githubusercontent.com/5993909/143213134-675fecdf-8ced-464a-a3a5-425897d9fe71.png)
![image](https://user-images.githubusercontent.com/5993909/143213581-caf334d7-0c61-4c8a-a49f-48d523965e7b.png)

Does it have something to do with the externals?

Edit: Yes. I have to use `externalsType: "module"` even though the external is not a module?

**Reply by JonWallsten** [reply]
I finally got all the pieces of the puzzle together. It works now!

However, I had to remove EvalSourceMapDevToolPlugin. 
Can this not be used with ESM?
![image](https://user-images.githubusercontent.com/5993909/143217928-10372ead-dc7c-45c8-84f2-27e8ab67c393.png)

**Reply by JonWallsten** [reply]
Doesn't seem to work with `devtool: 'eval-source-map'` either.

**Reply by alexander-akait** [reply]
Do you bundle library using `EvalSourceMapDevToolPlugin`? If yes, it should not work, we can't rewrite code inside eval, it is limitation

**Reply by alexander-akait** [reply]
> Edit: Yes. I have to use externalsType: "module" even though the external is not a module?

I think couple lines of code will say more https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L212

**Reply by JonWallsten** [reply]
I tried to bundle application only, but it was the same issue.
I really liked the `EvalSourceMapDevToolPlugin` because you could use custom `moduleFilenameTemplate` to get nice urls for each package in the monorepo. Like this:
``` 
app-1://[path]
app-2://[path]
lib-1://[path]
lib-2://[path]
```
Also had some custom logic for our Angular lib.
```ts
new EvalSourceMapDevToolPlugin({
    moduleFilenameTemplate: (info: any) => {
        // Match sourcemaps from Angular library
        if (info.resourcePath.match(/ng:[\\\/]{2}/)) {
            return info.resourcePath.replace(/ng:[\\\/]{2}@oas[\\\/]web-lib-angular[\\\/]/, 'web-lib-angular://./src/');
        }
        return `web-app-doc://${info.resourcePath}`;
    },
    exclude: /\.html|\.css|\.less|\.woff|\.woff2|\.svg|\.ttf|\.eot|\.jpg|\.png|\.gif|\.json|node_modules/
})
```
Is there any alternative? Will it work in the future?

**Reply by alexander-akait** [reply]
No alternative, you can't just do it, we can't rewrite code in eval, sorry

**Reply by JonWallsten** [reply]
No worries. Thanks for your help!
